supervisor/supervisor.go

```diff
@@ -0,0 +1,314 @@
+func (s *Supervisor) run(id string, proc *Process, sc <-chan *Status) {
+	go func(id string, p *Process) {
+		for {
+			select {
+			case l := <-p.StdOut():
+				line := LogOutput{StdOut, l}
+
+				s.access.Lock()
+				s.ouputBuffers[id] = append(s.ouputBuffers[id], line)
+				watchers := s.outputChannels[id]
+				s.access.Unlock()
+
+				watchers.sendLog(line)
+			case l := <-p.StdErr():
+				line := LogOutput{StdErr, l}
+
+				s.access.Lock()
+				s.ouputBuffers[id] = append(s.ouputBuffers[id], line)
+				watchers := s.outputChannels[id]
+				s.access.Unlock()
+
+				watchers.sendLog(line)
+			case <-p.Done():
+				return
+			}
+
+		}
+	}(id, proc)
+
+	fs := <-sc
+
+	if fs.ExitCode != 0 || fs.Error != nil {
+		// process didn't exit successfully. Log the error
+		s.logger.Error("Process didn't exit successfully",
+			zap.String("id", id),
+			zap.Int("exit code", fs.ExitCode),
+			zap.Error(fs.Error))
+	}
+}
+
+// Wait returns the done channel from Process. This channel is closed when the process stops running
+func (s *Supervisor) Wait(id string) (chan struct{}, error) {
+	s.access.Lock()
+	defer s.access.Unlock()
+
+	p, ok := s.jobs.Load(id)
+	job := p.(*Process)
+	if !ok {
+		s.logger.Error("Called wait on a job that does not exist", zap.String("id", id))
+		return nil, ErrJobNotFound
+	}
+
+	return job.Done(), nil
+}
+
+// WaitAll waits for all the processes to finish running
+func (s *Supervisor) WaitAll() {
+	s.wg.Wait()
+}
+
+// Stop stops a job by sending it's process group a SIGTERM signal
+func (s *Supervisor) Stop(id string) (*JobStatus, error) {
+	s.access.Lock()
+	defer s.access.Unlock()
+
+	p, ok := s.jobs.Load(id)
+	job := p.(*Process)
+
+	if !ok {
+		s.logger.Error("Cannot find job with the given id", zap.String("id", id))
+		return nil, ErrJobNotFound
+	}
+
+	if err := job.Stop(true); err != nil {
+		s.logger.Error("Cannot stop process", zap.String("id", id))
+		return nil, errors.Wrap(err, "cannot stop process")
+	}
+
+	// Wait for process to exit completely
+	<-job.Done()
+
+	// return final status to caller
+	st := job.GetStatus()
+	return &JobStatus{id, st}, nil
+}
+
+// StopAll stops a the jobs in the system
+func (s *Supervisor) StopAll() {
+	s.jobs.Range(func(id, j interface{}) bool {
+		job := j.(*Process)
+
+		if err := job.Stop(false); err != nil {
+			s.logger.Error("Cannot stop process", zap.String("id", id.(string)))
+		}
+
+		return true
+	})
+}
+
+// Status returns the current status of a job
+func (s *Supervisor) Status(id string) (*JobStatus, error) {
+	p, ok := s.jobs.Load(id)
+	job := p.(*Process)
+
+	if !ok {
+		s.logger.Error("Cannot find job with the given id", zap.String("id", id))
+		return nil, ErrJobNotFound
+	}
+
+	st := job.GetStatus()
+
+	return &JobStatus{id, st}, nil
+}
+
+// GetOutput returns the line buffer for the job
+func (s *Supervisor) GetOutput(id string) ([]LogOutput, error) {
+	s.access.Lock()
+	s.access.Unlock()
```

Comment:
defer?

Reply:
This been fixed. Don't know how I missed this